File: baekjoon/Java/1874_스택 수열.java
import java.util.Scanner;
import java.util.Stack;

public class P_1874 {
    static int num = 1;
    static boolean res = true;
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt(); // 수열 길이
        int[] arr = new int[N];
        for(int i=0; i <N; i++){ // 수열 입력받기
            arr[i] = sc.nextInt();
        }

        Stack<Integer> st= new Stack<>();
        for(int i=0; i<N; i++){
            if(arr[i] >= num) { // 만들어야하는 수열숫자 > 현재자연수
                while(arr[i]>=num){ // 현재 수열값과 값이 같아질때까지 push()
                    st.push(num++);
                    sb.append("+\n");
                }
                //값이 같아졌다면 pop()
                st.pop();
                sb.append("-\n");
            }
            else{ // 만들어야하는 수열숫자 < 현재자연수
                int n = st.pop();
                if(n != arr[i]){ // pop()했는데 값이 같지않으면 수열출력불가능
                    System.out.println("NO");
                    res = false;
                    break;
                }else{
                    sb.append("-\n");
                }
            }
        }
        if (res) {
            System.out.println(sb.toString());
        }
    }
}

File: baekjoon/Python/1874_스택 수열.py
import sys

input = sys.stdin.readline

N = int(input())
arr = [int(input()) for i in range(N)]
st = []
ans = []
num = 1
res = True

for i in range(N):
    if arr[i] >= num:  # 주어진수열값 >= 자연수
        # 숫자같아질때까지 스택쌓기
        while arr[i] >= num:
            st.append(num)
            num += 1
            ans.append("+")
        st.pop()  # 값이 같아지면 pop()
        ans.append("-")

    else:  # 주어진 수열값 < 자연수
        n = st.pop()
        if n != arr[i]:
            print("NO")
            res = False
            break
        else:
            ans.append("-")
if res:
    for op in ans:
        print(op)
